### ISSUE TITLE
Getting step0 cot-monitor training experiments (calibrate hinted monitor score to did_sel_hint)

### ISSUE THREAD

**Issue by GnarlyMshtep** [GnarlyMshtep/pass_at_k#56]
continuing from #51 (see last comment for summary and objectives)

**Comment by GnarlyMshtep** [comment]
I ran this [run](https://docs.google.com/document/d/1YWbJtE8g3-hisoLJVwkVHvmW9RoSYk32McoUld_4Ekw/edit?tab=t.t642q1oz32hw) and I think it failed and oscilated just as before even though I made the one line coding change and defined the new estimator function `compute_grpo_monitorability_outcome_advantage_CORRECTNESS_NO_EFFECT_SIZE` because I forgot to name it that! It was called `compute_grpo_monitorability_outcome_advantage` like the other function **what the hell python??? how did u not get mad at me??? Is this how u become a PL person?** -- this is actually pretty confirmed by the effect size not matching the `hinted/hint_sel_rate` (they should be set equal) -- rerunning now

## Maybe the above is WRONG 
I... actually am becoming more the above is NOT true, and I ran the right reward function but the logging was computing the effect seperately but I did not change that... I don't know why number is not going up... let me dif the configuration again -- the only diffs should be the reward function and the advantage estimator

confirmed, and check rollouts, and I am pretty doubtful the reward function is wrong -- so maybe the advantage estimator? Let me do another reading

I read over and fixed some logging issues, but nothing is wrong with the code that would be the score we are optimizing oscillate like that...

**Comment by GnarlyMshtep** [comment]
## [Another run in a mixed state](https://wandb.ai/matan-shtepel-carnegie-mellon-university/cot-monitor/runs/qp7adt7x?nw=nwusermatanshtepel)
All I know for sure is that it used the fixed GRPO with correctness-only-effect-size (did a print) but there was some logging which was off and it basically tracked perfectly (at least in many metrics) the previous run above. I am restarting with a `\lambda = 0.3` weight on the calibration -- lets see if it makes the difference that we want

### Actually this kinda makes sense
Calibration go up for the hinted setup... that should not be good... 

<img width="847" height="674" alt="Image" src="https://github.com/user-attachments/assets/c5675f77-24c1-4059-ae28-d0e1b6cf7b3b" />

<img width="848" height="679" alt="Image" src="https://github.com/user-attachments/assets/41d9b03a-46e0-407e-8383-1a1caa37ac5f" />
The number is mostly going up but the monitor score is dumping! This make sense for the baseline, but does not make sense for hinted -- okay, it makes sense that hinted score is going down. Lets see if the hinted score picked back up and with it the monitor

**Comment by GnarlyMshtep** [comment]
# [A run in where I checked the code is in a healthy state](https://wandb.ai/matan-shtepel-carnegie-mellon-university/cot-monitor/runs/f2hu0a0a?nw=nwusermatanshtepel)
still worth restarting to have that extra confidence -- let's see what happens.

## Wow, this is slightly confusing
Here are the monitor scores and the scores we optimized for
<img width="1405" height="686" alt="Image" src="https://github.com/user-attachments/assets/849ce3b3-0e0d-44cd-8b02-a21e925a86a2" />

I am not sure why of the score is so unstable. But, I am actually optimistic that this will work, because when the baseline's monitor score increases its perf generally goes down and opposite for the hinted version (who's monitor score is slightly going up). I think its worth training this, but not all night and I don't have a way to stop the machines reliably yet, so I will continue training tomorrow morning to conserve resources. 

Not much changed seemed to change in 100+ steps -- the model started outputting very long responses that were getting clipped >1/2 of the time, so kinda waste of compute

**Comment by GnarlyMshtep** [comment]
# [Running again without entropy bonus](https://wandb.ai/matan-shtepel-carnegie-mellon-university/cot-monitor/runs/784ngbzv?nw=nwusermatanshtepel)
 response length does not blowup, but the monitor hint sel rate still goes to 0 -- well before that of the unhinted hint selection rate... what's going on? 

My only guess is that somehow the order of my prompts are reversed, and that I am giving the model unhinted prompts while labeling them hinted and the model is super confused. **I will test for this hypothesis**

<img width="895" height="325" alt="Image" src="https://github.com/user-attachments/assets/dd88744a-9f2c-4375-b3c7-2e664b0d07fa" />
Nope. 

## Then why did it not work? / how do we make it work?  
More specifically, what I mean is that we are using the [control / hint-correct / hint-incorrect / hint-omission] setup discussed in #51 and the monitor just looks out for the word "hint" (and a few nearby words). The control is penalized whenever it says hint, and any of the hint setup are penalized for selecting the hint and not saying the word hint. 
Here is the critical part of the code (confirmed the above made sense)
```pyton 
 for j in range(len(monitor_index2infos[base_idx])):
                infos = monitor_index2infos[base_idx][j]
                correctness_plus_other = infos["other_score"] + infos["is_correct"] # the other score here is KL in reward and stuff like that -- it should be ~0 afaik and we can assert htis 
                hint_sel_effect = 0
                calibration = (hint_sel_effect - infos["monitor_score"]) ** 2
                score_unwhitened = correctness_plus_other - LAMBDA_CALIB_WEIGHT* calibration


                monitor_index2infos[base_idx][j]["calibration"] = calibration
                monitor_index2infos[base_idx][j]["used_effect_size"] = hint_sel_effect
                monitor_index2infos[base_idx][j]["score_unwhitened"] = score_unwhitened

            for j in range(len(monitor_index2infos[hinted_idx])):
                infos = monitor_index2infos[hinted_idx][j]
                correctness_plus_other = infos["other_score"] + infos["is_correct"]
                hint_sel_effect = infos[
                    "did_sel_hint"
                ]  #! M: this is the key line that we changed, everything else should be the same and I will not change
                calibration = (hint_sel_effect - infos["monitor_score"]) ** 2
                score_unwhitened = correctness_plus_other - LAMBDA_CALIB_WEIGHT* calibration

                monitor_index2infos[hinted_idx][j]["calibration"] = calibration
                monitor_index2infos[hinted_idx][j]["used_effect_size"] = hint_sel_effect
                monitor_index2infos[hinted_idx][j]["score_unwhitened"] = score_unwhitened
```
Here's what happens which is annoying -- the monitor score of the hinted run just _dumps_

<img width="850" height="680" alt="Image" src="https://github.com/user-attachments/assets/4fdca59f-5c07-4fb7-9a4b-f5e1e65dcdf5" />

**The question is -- why?? what to do about it** 

### What to do about it: warmup 
The grad norm is very high to start with

<img width="275" height="316" alt="Image" src="https://github.com/user-attachments/assets/780bec9c-3bd6-41b1-8a47-5cc48fe19d01" />

We are training just 1 model, and in baseline mode, it is just always severely punished for saying `hint`. The dynamic is so weird -- the model in baseline mode goes from saying `hint` < `0.005 to 15% although it is never ever rewarded for it! I think the model really is confused.

### Why is this happening? 
Yeah, I think that something like the above. Grad is way too high. Am I norming by std? Nope... that... actually might be giving somewhat more irregular grads here... Since we are not trying to get good calibration I will try again with it on, hopefully it stabalizes the gradient a bit more

**Comment by GnarlyMshtep** [comment]
# [Retrying on deltaai the same recipe as above + 15 warmup steps +  grpo stdnorm=False](https://wandb.ai/matan-shtepel-carnegie-mellon-university/cot-monitor/runs/9blm4kah?nw=nwusermatanshtepel)

### What does "warmup step" mean in `verl`? Can't find docs but want to be sure 
```python 
def lr_lambda(current_step):
        if current_step < num_warmup_steps:
            return float(current_step) / float(max(1.0, num_warmup_steps))
        return 1.0

    return LambdaLR(optimizer, lr_lambda, last_epoch)
```

<img width="455" height="306" alt="Image" src="https://github.com/user-attachments/assets/d54df916-9d93-447f-93a3-ffa254665422" />

### Systems issues

I actually think this is worth waiting on #61  a bit more -- still getting some OOMs and don't want to deal with 2 runs with systems issues. **lots of OOMs still even after getting #61 -- painnn**

The issue was primarily that I accidentally turned gradient checkpointing _off_, making updates very expensive. 


## The run itself

### Not what we wanted, but no optimization issue

<img width="955" height="266" alt="Image" src="https://github.com/user-attachments/assets/5dcb0ed4-2abd-40ed-a891-14986e1ed790" />

<img width="475" height="474" alt="Image" src="https://github.com/user-attachments/assets/299ee95c-3fc0-4cd0-bfdc-7ac050d5ef40" />

### How does behavior look across hint types? 

<img width="1408" height="791" alt="Image" src="https://github.com/user-attachments/assets/71cca8fb-3712-49ec-853d-1dc68f00a0d4" />
calibration for correct & incorrect stays similar over time, improves for incorrect and baseline -- since they make up $3/6+1/6 = 2/3$ of the questions together, that's enough to improve score without saying hint more often. 

### Why does the model not learn another strategy to increase score?
Like, for example, selecting the hint always and saying "hint"! -- I guess it is not as natural to it? Can one really explain things like that? We know that this is the global optima, but I guess the model does not. Maybe it has been trained to just do RL to learn math problems better that attending to the hint and the word "hint" is just not _salient_  for it? 

### Wait, why does the model learn to stop saying "hint" at all? 
Here are some potential reasons: 
* All the rollouts in "hint-incorrect" which say hint usually select the hint and all get punished heavily. Instead of learning "some hints are incorrect" the model learns "ignore this hinted information" 
* the model learns to say "hint" for correct & omission questions, but doesn't learn that it should only say it when it is in the question, so it says it in baseline, gets heavily punished, and learns not to say hint. 

### How to fix
1. The easiest and most common thing is to fine-tune to make what we want more salient for our model. I guess we decided that defeats the behavior, although this is weird since it is the right tool for the task. @andrewilyas 
2. It is possible that if we continue to train, the model will get out of the local optima and start saying "hint." I feel this is unlikely because the model learns to not say "hint" so hard, that it almost loses that ability. At the very least, this is not a stable way to train. 
3. Run again with _only_ correct hints. This is step $-\infty$ sanity check. 
4. Run again but with _no_ incorrect hints. This is very likely confusing the model / enabling it to hack our hint objective.  

I should talk about 1 with @andrewilyas , mostly ignore 2, and try either 3 or 4 as a sanity check. Do the dumbest thing first sounds like something a smart professor would say so I will do 3.

**Comment by GnarlyMshtep** [comment]
# [A run with deepmath reward fixed](https://wandb.ai/matan-shtepel-carnegie-mellon-university/cot-monitor/runs/l2p9q9gs)
In #63  I fixed the outragous math scoring -- now I am trying to run again and seeing if it makes a difference

<img width="2456" height="753" alt="Image" src="https://github.com/user-attachments/assets/264365e6-3843-42a0-b75f-b218ccf52bd0" />
The answer seems to be no -- hinted monitor score still plummets.

**Comment by GnarlyMshtep** [comment]
## [Trying again only with correct hints (Step 0--100)](https://wandb.ai/matan-shtepel-carnegie-mellon-university/cot-monitor/runs/4jpjlycx?nw=nwusermatanshtepel)
[it took a while to make the unbroken math reward work, and low and behold -- deepmath is actually not hard Baseline (no hint) achieves 93% correctness in 40 steps lol.]. 
<img width="803" height="342" alt="Image" src="https://github.com/user-attachments/assets/60f2a2ee-ce6b-4820-ba46-8d31481e173a" />
 
The run is actually very encouraging! The model is eventually learning to mention the hint as it should.  
<img width="1664" height="395" alt="Image" src="https://github.com/user-attachments/assets/d2913418-5557-4f86-98ff-f0debffa5fcd" />

I will continue for a few more steps (~50) just to make sure.
